docs/adr/2022-07-22_019-reference-scripts.md

```diff
@@ -0,0 +1,61 @@
+---
+slug: 19
+title: | 
+  19. Use of reference scripts
+authors: []
+tags: [Proposed]
+---
+
+## Status
+
+Proposed
+
+## Context
+
+* In the desire to make Hydra transactions smaller and cheaper (at the time of writing any abort tx was too big), we want to use the **reference script** and **reference input** features of the upcoming `Babbage` ledger era. See the [babbage ledger spec](https://hydra.iohk.io/build/16861604/download/1/babbage-changes.pdf), [CIP-31](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0031) and [CIP-33](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0033) for details.
+
+* With these features we do not need to (re-)include scripts in each transaction.
+
+* The CIPs do not specify how reference scripts are to be managed and we can see at least two options:
+  1. Add them as outputs to the `init` transaction or prior that as part of each Hydra Head instance
+  2. Post them out-of-band, separate to individual Head instances
+
+* Ownership of the outputs holding the scripts is to be considered. If these "reference outputs" are spent, they cannot be referred to anymore. This would mean all heads referring to them can be denied of service (DoS).
+
+* Each head will need to refer to the correct version of the hydra scripts. That is, consistent with the script hashes known to the `hydra-node`.
+  + This is also related to the problem of managing script versions & updates.
+  + Right now, the `hydra-node` is compiled against `hydra-plutus` to access compiled script content and hashes.
+
+* The general trade-off is: instead of paying ADA fees for scripts adding to the transaction size in _each_ transaction, ADA deposits will need to be put down to have scripts be part of the UTxO set in the ledger _once_.
+
+## Decision
+
+* Publish outputs holding Hydra scripts out-of-band (option 2), because
+  + All scripts would not fit into the `init` transaction directly, we would need to post multiple.
+  + Costs (deposits) would need to be payed for each head instance.
+  + Although, heads using option 1 would not be exposed to DoS 
```

Comment:
The way this is presented is a bit weird because, both option (1) and option (2) are about publishing Hydra scripts on-chain; the only question is whether or not to mix that with the init transaction. 

And the conclusion being that no, we'd rather not for reason X & Y. 

>    + Although, heads using option 1 would not be exposed to DoS 

I disagree, it would be exposed to it all-the-same. The DoS problem stems from what locks the outputs where the script is posted to. If it's spendable, then it's deniable. So these are kind of orthogonal. 

One extra reason why doing it as part of the init isn't a good idea is the complexity it adds to the system. When the publishing transaction is _assumed_ to exist, it is known on application start, provided as configuration. However, doing it part of the init means that there needs to be code that (a) realize the event and (b) monitor the outcome of the event. That's extra complexity for the Hydra node which we may want to prevent.

Reply:
Good points. You are right in that the DoS stuff is about spendability of the outputs. In option 1 we could make it spendable by head instance itself and only when terminating it, i.e. by ensuring the ST is also burned. So it can be made DoS-safe and non-littering. Which is not possible in option 2.